src/Umbraco.Core/PropertyEditors/IDataValueEditor.cs

```diff
@@ -35,17 +35,20 @@ public interface IDataValueEditor
         bool HideLabel { get; }
 
         /// <summary>
-        /// Gets a value indicating whether the value can be copied
+        /// Gets a value indicating whether the value can be copied.
         /// </summary>
         bool CanCopy { get; }
 
         /// <summary>
-        /// Validates a property value.
+        /// Validates a property value using custom messages.
+
         /// </summary>
         /// <param name="value">The property value.</param>
         /// <param name="required">A value indicating whether the property value is required.</param>
+        /// <param name="requiredMessage">A custom validation message to use when the property value is required.</param>
         /// <param name="format">A specific format (regex) that the property value must respect.</param>
-        IEnumerable<ValidationResult> Validate(object value, bool required, string format);
+        /// <param name="formatMessage">A custom validation message to use when the property value is does not match the specific format (regex).</param>
+        IEnumerable<ValidationResult> Validate(object value, bool required, string requiredMessage, string format, string formatMessage);
```

Comment:
I'll have a think... though I did notice in getting this PR up to date, that this interface was changed with the addition of `CanCopy` property about 3 weeks ago, so this breaking change issue may existing with minor upgrades already.

Would adding a new method (i.e. an overload) be better than changing the existing signature do you think?  Or is that still as much of a concern?

Reply:
Hi @AndyButland i think that change somehow slipped through the cracks, we're gonna have to adjust that somehow before the next release too so there's no breaks before 8.2.

The problem with not changing the interface is that you are then gonna have to do a type check on the concrete type and then call that method which is a little uglier than having a version 2 of the interface since at least that allows you to mock stuff. Another alternative is to have just another standalone interface containing the changes but that's also not exactly nice.

We'll have a chat about these changes, perhaps we can allow breaks for this in 8.3 but will let you know soon :)

Reply:
OK, will leave for now then - as if you take away the breaking changes issue, then of course it's cleaner as is, to just change the interface.  But if that's not feasible let me know, probably along with which of the suggestions you've made you think best, and I can look to amend.

Reply:
Hi @AndyButland , I 'think' i might have a way to acheive this without breaking changes. 

The underlying mechanism that calls the validation is only in 2 places: 

* ContentModelValidator.ValidatePropertyValue
* PropertyValidationService.IsPropertyValueValid

I 'think' in these 2 places we could call into `IEnumerable<ValidationResult> Validate(object value, bool required, string format);` 2 different times depending on the values there are for `required` and `format`. So first we would check if the property type is marked required, if so, we would call this method with `required: true, format: null` and get the response, if it was invalid then we can assign the property type's required message to the outgoing validation. Then we could call this method again if the property type had a regex format applied and call the method like `required: false, format: format` and do the same again, if the result was invalid we'd assign the property type's format message to the outgoing validation. 

From what i can tell this should be possible and then none of these interfaces need to change, it is then just the responsibility of the callers to the `Validate(object value, bool required, string format);` method to either use the error message returned or replace it with a configured one in the property type.

Does that make sense? Maybe it's a little ugly but this could then all be wrapped in an extension method on `IDataValueEditor`

Reply:
Thanks for the pointers, I'll take a look later on and see if I can make some progress.

Reply:
Unfortunately I ran into an issue with this @Shazwazza  - which is that the implementation of `DataValidator.Validate` will check not just required and regex validation but also any custom validators that are associated with the property type, `UploadFileTypeValidator` being one example.

So there actually isn't a way, with the current method signature, to tell it to just do required validation, or regex validation... we'll do one or other of those, but also any custom validations too.  And as such any `ValidationResult` objects in the collection that's returned, may be from the required/regex validation we've requested, or may come from the custom validators.  We'll also do the custom validators twice, potentially ending up with duplicate validation errors.

So I tried a different tack.  Given we know what the default message applied will be, we can just look for the cases where we have a `ValidationResult` with those error messages, and replace them.

Seems we actually only need this in `ContentModelValidator`.  The other instance you noted, in `PropertyValidationService` is just interested in a valid or not boolean results, so we don't care about messages there.  Given that, didn't see there was really a need to pull this out into an extension method, it can just live in `ContentModelValidator`.

So I've done that in the latest commit - as well as reverting the interface changes.  Please let me know how that looks to you.  It's still a little ugly I think relying on matching error messages - and I suppose there's a very small edge case in a custom validator happening to use the same message string -  but it looks to do the job.

Reply:
Thanks @AndyButland ! It is a bit ugly so i guess now we need to decide: Do we allow breaking changes and so this the right way or do we do this the hacky way. There are some other upcoming changes to property editors with regards to media tracking that may end up resulting in a requirement to change interfaces (not decided yet), so maybe if that happens we can align these breaking changes in the same release. 

I have discovered that we are already (unfortunately) casting `IDataValueEditor` -> `DataValueEditor` 100% of the time within `DataEditor.GetValueEditor` and there is a note: `// TODO: casting is bad` which i assume used to be a `fixme` meaning we probably didn't want to ship with this. So in theory, the `Configuration` should be part of the interface too so we don't cast. 

Another alternative to what you've done would be to:

* Obsolete the `IDataValueEditor.Validate` method ... it could be left as-is though
* Update the `ValidatePropertyValue` method:
```cs
protected virtual void ValidatePropertyValue(
    TModelSave model,
    TModelWithProperties modelWithProperties,
    IDataEditor editor,
    ContentPropertyDto property,
    object postedValue,
    ModelStateDictionary modelState)
{
    // validate
    var valueEditor = editor.GetValueEditor(property.DataType.Configuration);

    // validate each custom validator
    List<ValidationResult> results = null;
    var r = valueEditor.Validators.SelectMany(v => v.Validate(postedValue, valueEditor.ValueType, property.DataType.Configuration)).ToList();
    if (r.Any()) { results = r; }

    // TODO: Casting is bad :(
    if (valueEditor is DataValueEditor dataValueEditor)
    {
        // validate required/regex
        if (property.IsRequired)
        {
            r = dataValueEditor.RequiredValidator.ValidateRequired(postedValue, valueEditor.ValueType).ToList();
            if (r.Any()) { if (results == null) results = r; else results.AddRange(r); }
        }

        var stringValue = postedValue?.ToString();
        if (!string.IsNullOrWhiteSpace(property.ValidationRegExp) && !string.IsNullOrWhiteSpace(stringValue))
        {
            r = dataValueEditor.FormatValidator.ValidateFormat(postedValue, valueEditor.ValueType, property.ValidationRegExp).ToList();
            if (r.Any()) { if (results == null) results = r; else results.AddRange(r); }
        }
    }
            
    foreach (var valResult in results)
    {
        modelState.AddPropertyError(valResult, property.Alias, property.Culture);
    }
}
```

Essentially just moving the full validation logic here which includes checking/casting which isn't nice but i think it might be nicer for now than checking validation message responses. Then from there we can decide when/how to break the interfaces to do all of this correctly. 

What do you think?

Reply:
I'm not quite following where in this last suggestion we'd be setting the custom messages though?  As if we do that in `RequiredValidator` and `RegexValidator`, we'd need to pass in the message to use, which would involve a change to `IValueRequiredValidator` and `IValueFormatValidator` interfaces.  And if we do it in `ValidatePropertyValue`, then seems we have to do the checking for matches on the error messages still.

Reply:
@AndyButland ah yes of course since the result of ValidateRequired and ValidateFormat return multiple errors, we'd have to do the error message check anyways. Lets stick to your current solution for now, maybe that's just the simplest way to do it for now. Perhaps we should `Contains` to `ContainsInvariant` though just to be safer? Also just want to confirm this would still work when using a non-default back office language?

Reply:
Have updated to use `StringComparer.OrdinalIgnoreCase`.  I haven't tested yet using a non-English back-office language (and am away for a few days, so won't be able to do so shortly), but I believe it would work as expected.  We're using the localisation text service to get the default validation messages, which uses the configured culture, so it means we'd have the localised ones set and used for the check.

Reply:
Confirmed this works OK when I use a different back-office language and ensure these default validation messages are added to the appropriate language file.

Reply:
Great stuff, i think this is in a good place, i'll find out from the team which version we should release this in and get the ball rolling. Thanks @AndyButland !!